### ISSUE TITLE
Export to plain HTML

### ISSUE THREAD

**Issue by mrgou** [giuspen/cherrytree#1166]
I would love to see an option to export to plain HTML, i.e. a single file (with references to external files such as images) in plain HTML, where node titles are included in nested heading elements (`H1`, `H2`, `H3`...), lists in standard `ol` and `ul` elements...

A use case for CherryTree for me is to help me author highly structured documents, such as procedures, manuals, etc. Contrary to a regular word processor, having just one text page for each section allows focusing on that one only. It is then also very easy to reorganize the structure of the document.

But then, for the final step, it needs to be exported to a word processor for formatting and publishing, so a plain HTML file would be a good transition format.

**Comment by ciranus** [comment]
> I would love to see an option to export to plain HTML, i.e. a single file (with references to external files such as images) in plain HTML

Is **index.html** not the file you are looking for ?
(or not sure to understand what you mean by "plain html").

**Comment by funnelfiasco** [comment]
I think @mrgou is asking for a single-file output. Given a node "top node" that contains children "childa" and "childb", the HTML output would be something like

    <html>
    <title>top node</title>
    <h1>childa</h1>
    Contents of node childa
    <h1>childb</h1>
    Contents of node childb
    </html>

Is that right, @mrgou ?

**Comment by ciranus** [comment]
> I think @mrgou is asking for a single-file output. 

Yes but ... "with references to **external** files such as images".
Is "_such as images_" only images or anything else ?
If the data-base is a 100MB sqlite file, is is a good idea to have a single 100MB html file ?
To be clarified.

**Comment by mrgou** [comment]
@funnelfiasco is correct, just like the existing Single Plain Text File option. Ideally, there should be references to images only, as they can't be embedded in HTML, and no interactive code. And I suppose the file could be very large (but it would take a lot of content for that). The idea is to facilitate import in a word processor.

**Comment by ciranus** [comment]
> images only, as **they can't be embedded** in HTML

Images can be embedded in HTML (converted to `data:image/png;base64` format).
`echo "data:image/png;base64,$(base64 -w 0 image.png)"` for Linux.
HMTL image code: `<img src="data:image/jpeg;base64,"DFGQDFGQDFGDFHSFH...">`

If all files are not included, the benefit of your request is limited (why external images, and not external HTML nodes, the main page being index.html).
What you are looking for is same as [SingleFile](https://addons.mozilla.org/en-US/firefox/addon/single-file/) Firefox addon (very good addon).
Your request is then either:

1 - an extra "Export as single HTML file" option (all included) , or
2 - an extra "Export as single HTML file" option (but with external images to make the HTML code faster to load).

**Comment by giuspen** [comment]
I think the single plain html page with the images embedded as base64 could be useful as an alternative to the pdf. It would in fact be like the .ctd (xml) with the benefit of being opened in read more from any web browser.

**Comment by ciranus** [comment]
Agreed.
PDF is too limited to make something nice.
But this should be an extra option, not replacing current one.
The problem of single file is the size.
Browsers do not appreciate 100MB HTML files.....
And easier to edit small clearly identified HTML node files, than a single big one full of image data code.....

**Comment by mrgou** [comment]
Yes, images can be embedded in HTML, but that may not be the best approach--I leave it to you to decide. In the end, a word processor like LibreOffice writer will import external referenced images into a document, but not other HTML files to which there is a link, like in the current `index.html`. They can also handle large files.

**Comment by ciranus** [comment]
@mrgou
As you see, all combinations are possible. There is no technical limitation.
If you now seem to consider that external images is probably better, why do you explicitly consider that external HTML node files is a problem ?
The "SingleFile" concept of Firefox addon is to create something easily portable, like a PDF single page, but improved with HTML capabilities.

> I leave it to you to decide.

Don't you need more time to clarify your real need ?

**Comment by mrgou** [comment]
> If you now seem to consider that external images is probably better, why do you explicitly consider that external HTML node files is a problem ?

Simply because they won't be imported and inserted into a single document when importing the top-level HTML file (`index.html`) in a word processor, contrary to external images. There's a big difference between a link to another HTML file and an image reference.

**Comment by ciranus** [comment]
> Simply because they won't be imported and inserted into a single document when importing the top-level HTML file (`index.html`) in a word processor, contrary to external images.

This clarifies that your goal is to edit HTML in a "word" processor (and not making an easy portable document).
Why your word processor is importing images ? Not a generic rule at all.
If I use kate which is a word editor, no image will be imported, but that's clear that all HTML nodes will never be imported in one document, you just have to edit the specific node document you need to edit, and which is clearly identified by its name.

SingleFile will not look for any external html pages:

> SingleFile is an add-on for Firefox that helps you to save an **entire webpage including images, styling, frames, fonts etc. as a single HTML file**.

Then in summary, if this option justified or not:

> 2 - an extra "Export as single HTML file" option (but with external images to make the HTML code faster to load).

It's not up to me to decide, but all is now detailed and clarified.

**Comment by mrgou** [comment]
> This clarifies that your goal is to edit HTML in a "word" processor (and not making an easy portable document).

Yes, as I explained above:

> A use case for CherryTree for me is to help me author highly structured documents, such as procedures, manuals, etc. Contrary to a regular word processor, having just one text page for each section allows focusing on that one only. It is then also very easy to reorganize the structure of the document.
> But then, for the final step, it needs to be exported to a word processor for formatting and publishing, so a plain HTML file would be a good transition format.

> Why your word processor is importing images ? Not a generic rule at all.

I can't say in general, but LibreOffice Writer and MS Word do import images referenced in an HTML file.

**Comment by ciranus** [comment]
> LibreOffice Writer and MS Word

Word tools are not ideal anyway to edit HTML.
How do you add a node (if necessary) with a word tool from CT HTML export ?
I know how to do it with a text editor, but not with word tools.
And how do you modify the css stylesheet, if necessary ?
+ new question: css stylesheet : integrated or not ?

Your editing capabilities are then very limited with word tools.
Frankly not convinced that the single plain HTML file will facilitate you anything, but this is just an only my personal feeling, which should not interact with any decision.

**Comment by txe** [comment]
@mrgou, I see your point, I'm going to add this export option soon.

**Comment by mrgou** [comment]
> @mrgou, I see your point, I'm going to add this export option soon.

Yay!

**Comment by txe** [comment]
I added an option 'Single File' on an export dialog which you see after starting 'Export to HTML' command

**Comment by mrgou** [comment]
Thanks. Contrary to what I had in mind, tree levels are not mapped to HTML headers levels.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<cherrytree>
  <bookmarks list=""/>
  <node name="Level1" unique_id="1" prog_lang="custom-colors" tags="" readonly="0" custom_icon_id="0" is_bold="0" foreground="" ts_creation="1601311145" ts_lastsave="1601311147">
    <rich_text>Text</rich_text>
    <node name="Level2" unique_id="2" prog_lang="custom-colors" tags="" readonly="0" custom_icon_id="0" is_bold="0" foreground="" ts_creation="1601311209" ts_lastsave="1601311211">
      <rich_text>Text2</rich_text>
    </node>
  </node>
</cherrytree>
```
results in:

```html
<!doctype html><html>
    <head>
      <meta http-equiv="content-type" content="text/html; charset=utf-8">
      <title>test.ctd</title>
      <meta name="generator" content="CherryTree">
      <link rel="stylesheet" href="res/styles3.css" type="text/css" />
    </head>
    <body>
    <div class='page'><h1 class='title'>Level1</h1><br/>Test</div><div class='page'><h1 class='title'>Level2</h1><br/>Text2</div></body></html>
```

I was hoping to see `Level2` wrapped in an `H2` element.

**Comment by txe** [comment]
@mrgou, sorry it took so long. 
I just tried your idea (changed html file in an editor). h2-h7 tags make titles smaller, and even smaller than an ordinary text. It looks ugly and doesn't show hierarchical tree structure at all.

**Comment by mrgou** [comment]
OK. What mattered to me in the initial idea was to maintain the tree structure when importing in a Word processor (where heading styles can then be adjusted). In HTML, display formatting can be tweaked with CSS, but the most important was the semantics of the headings and sub-headings.

**Comment by txe** [comment]
I can add some info (like `level-1`)  for titles or even for the whole nodes, so you can apply css, e.g.
`<h1 class='title level-1'>Level1</h1>`

**Comment by mrgou** [comment]
That wouldn't be interpreted as heading level by a Word processor, but I could still transform that `H1` element based on the class into a corresponding heading element.

**Comment by txe** [comment]
Fixed, FYI, it is not `title-level-1`, but `title level-1` (two separate classes)

**Comment by mrgou** [comment]
It works perfectly, thank you! :-)
